yb-voyager/src/tgtdb/yugabytedb.go

```diff
@@ -300,10 +300,46 @@ var NonRetryCopyErrors = []string{
 	SYNTAX_ERROR,
 }
 
+// isDataIntegrityOrConstraintError checks if an error is a data integrity or constraint violation or syntax error
+// by examining the SQLSTATE code.
+//
+// SQLSTATE Class 22: Data Exception (e.g., 22003=numeric overflow, 22P02=invalid syntax)
+// SQLSTATE Class 23: Integrity Constraint Violation (e.g., 23502=not null, 23505=unique)
+// Postgres Error Codes: https://www.postgresql.org/docs/current/errcodes-appendix.html
+func isDataIntegrityOrConstraintError(err error) bool {
```

Comment:
we might want to name this method a bit more generically, so that when we add more error classess, it naturally fits in.